various amendments

dornech opened the pull request:
Hi @FlorianWilhelm,
I spent some time on your Hatchlor. 
I did some amendments you left over: I included two options for local comitting and versioning - bump-my-version/generate-changelog or semantic-release. Both are - if set up - easy to use. I included a basic config for both tools in the pyproject.toml template so everbody is free to use the one or the other.
Furthermore I added gitlint which I find useful to enforce conventional commit compliant commit messages - pretty useful if you automate the changelog generation with generate_changelog or semantic-release. gitlint is cool stuff and for example works also within PyCharm which is my favorite IDE (which is probably not surprising for the more professional Python developer than I would claim for myself).
Finally I did some rework on the GitHub workflows.
These are the main changes I think. Please find approriate comments in the README.md as well.

dornech committed:
Feat: rework GitHub workflows

dornech committed:
Feat: add config files for linting of commit messages

dornech committed:
Feat: add config file for generate-changelog

dornech committed:
Feat: adjust pre-commit.yaml for linting commit-msg and version update

dornech committed:
Fix: correct build.yaml

dornech committed:
Fix: amend .gitignore for ruff cache

dornech committed:
Feat: include hatch-vcs-footgun-example in __init__.py

dornech committed:
Feat: various changes to pyproject.toml to include additional features

dornech committed:
Update README.md

dornech committed:
feat: code blocks in mkdocs.yaml

dornech committed:
feat: add. settings pymdownx.highlight

FlorianWilhelm commented on {{cookiecutter.project_slug}}/src/{{cookiecutter.pkg_name}}/__init__.py:
```diff
@@ -5,9 +5,32 @@
+    # own developed alternative variant to hatch-vcs-footgun overcoming problem of ignored setuptools_scm settings
+    # from hatch-based pyproject.toml libraries
+    from hatch.cli import hatch
+    from click.testing import CliRunner
```
What is the advantage of this compared to the original version? It expects way more packages to be installed during execution time like `click` and `hatch` itself, which normally is not required to be present when you just want to install the package you have developed.

dornech replied:
This is required for retrieving the actual build version. It is an adjusted version of the footgun as commented. 
I agree this might be overhead for non-developer installation but if hatch is installed anyway, there is no additional overhead.

FlorianWilhelm replied:
Not sure I really get it. My original code accomplished the same thing. It read out the version but only using modules that are shipped directly with Python. So the big advantage of my original code is that you don't need to have hatch installed to use the package. I know that developers of a package need to have hatch installed but you should always think about the users of what you are building. They should not be required to install hatch.

dornech replied:
If you do an "pip install -e" then you always see the last install version -- not the last commit version.
It is explained here:
https://github.com/maresb/hatch-vcs-footgun-example

But I agree: this is probably a small advantage compared to the setuptool_scm dependency. And it is only neceesary if one has very strict version dependencies in the code.
Probably refering to metadata or the local _version.py after a hatch build is a good compomise?

dornech replied:
Argh, I did miss that I eliminated the setuptools_scm dependency already because of using Hatch compared to the original footgun-version. the setuptools_scm did not work correctly with the version bumping.
Explained in more detail here - https://github.com/maresb/hatch-vcs-footgun-example/issues/3: @maresb made also an suggestion how to call hatchling directly (without the click-overhead of Hatch as the front-end to hatchling).

FlorianWilhelm commented on {{cookiecutter.project_slug}}/pyproject.toml:
```diff
@@ -71,10 +77,104 @@ version-file = "src/{{ cookiecutter.pkg_name }}/_version.py"
 packages = ["src/{{ cookiecutter.pkg_name }}"]
 
 [tool.hatch.build.targets.sdist]
+artifacts = ["_version.py"]
```
Smart, so this is necessary for source distributions so that it is included?

dornech replied:
The _verison.py file is generated during build and is to be included for the installable installation I think.

FlorianWilhelm replied:
This worked out of the box for wheel files but not sure about sdists. Good point.

dornech replied:
To be honest I am not experienced with the difference between wheel and sdist-distributions.

dornech replied:
However, what I see from my testpackage the _version.py is included in sdist and wheel distribution.

FlorianWilhelm commented on {{cookiecutter.project_slug}}/pyproject.toml:
```diff
@@ -26,20 +26,26 @@ authors = [
 # direct dependencies of this package, installed when users `pip install {{ cookiecutter.project_slug }}` later.
 dependencies = [ # ToDo: Modify according to your needs!
-    "typer",
+    "setuptools_scm",
```
setuptools_scm should not be needed for a normal, i.e. non-development, installation. It's not necessary to have it as a runtime requirement.

dornech replied:
I agree but it is the issue as above: it is the hatch-vcs-footgun in __init__.py ...

dornech replied:
Argh, I did miss that I eliminated the setuptools_scm dependency already because of using hatch compared to the original footgun-version. the setuptools_scm did not work correctly with the version bumping.

dornech replied:
BTW - the issue with setuptools_scm was discussed here 
https://github.com/maresb/hatch-vcs-footgun-example/issues/3

FlorianWilhelm commented on {{cookiecutter.project_slug}}/.pre-commit-config.yaml:
```diff
@@ -17,18 +17,25 @@ repos:
       args: ["--install-types", "--non-interactive"]
       additional_dependencies: [types-tabulate, types-cachetools]
+
+-   repo: https://github.com/jorisroovers/gitlint
```
I really like the idea but would it be better to have this optional via cookiecutter? Gitlint is still having less than 1k Github stars and is it really used that often? I like it but it's not a commonly found best practice, is it?

dornech replied:
Well, to be honest: gitlint was the only commit-message linter I found. For automated generation of a changelog it is necessary to have proper commit messages. I struggled somewhat on that ... 
Probably more professional developers than I am are much more familiar with this and do it while asleep but for me it was helpful.
I do not knoy how cookiecutter would be of any help here but probably I missed something there.

dornech replied:
Ah, I guess now I know what you mean with "optional via cookiecutter". You mean similar to the lock-file support i. e.  controlling via y/n prompt during cookiecutter execution? I am afraid it is me lacking some cookiecutter / Python packaging experience that I did not get it directly. Sounds like a valid option.

Probably also an option fot the complete footgun-stuff ...

FlorianWilhelm replied:
Exactly, to make it gitlint an additional option. Bear in mind though that with every additional option, running unittests becomes more complicated since you need to test every possible path, not only the one that includes all features.

dornech replied:
Question: to do it perfect, I would have to delete the .gitlint config file after generation via the post_gen_project.py hook ?

dornech replied:
I tried to make the gitlint optional. Struggling with the post-generation hook to delete the .gitlint file, other error fixed.

dornech replied:
Works now

FlorianWilhelm commented on {{cookiecutter.project_slug}}/.github/workflows/tests.yml:
```diff
@@ -38,4 +45,4 @@ jobs:
       - name: Run Tests
         run: |
-          hatch run test:pytest
+          hatch test
```
Nice, the new way of doing this! Thanks.

dornech replied:
My idea was to leave the testing for different versions and environments mainly with GitHub. This is achieved by setting the matrix for os /  python version in the GitHub action and executing a hatch test in the standard hatch-test environment.
I was only struggling a little bit with the package dependencies, there might be some room for optimization.

FlorianWilhelm reviewed (commented):
Thanks @dornech! I added a few comments.

Would you be interested in becoming a co-maintainer for The Hatchlor? 🌹

dornech committed:
fix: escaping of variable in .changelog-config.yml

dornech committed:
fix: escaping of variable in .changelog-config.yml /2.

dornech committed:
fix: escaping of semantic-release config in pyproject.toml

dornech committed:
fix: correct build.yml (incorrect escaping)

dornech committed:
fix: correct build.yml (incorrect escaping) /2.

dornech committed:
fix: correct build.yml (incorrect escaping) /3.

dornech committed:
fix: correct pyproject.toml (incomplete escaping)

dornech commented:
General remark: I am afraid this pull request still has some escaping errors. Fix see subsequent commits from today.

dornech commented:
Regarding co-maintainer: I feel honoured you are asking. I am just not sure if I am experienced enough. I mean I think I am not a rookie in IT and in development but at least a newbie in the Python Packaging universe ...

dornech committed:
fix: editing ...

dornech commented:
I would love to see PyScaffold become Hatch-enabled to allow to work with Hatch-based scaffolds and probably migrating Hatchlor to a PyScaffold template. So far I am struggling with Cruft which is less well maintained and does not work in my environment.

dornech committed:
feat: add basic Cruft config (template update helper)

FlorianWilhelm commented:
PyScaffold follows a completely different approach and uses traditional tools like setuptools only. To make it hatch-based would mean to rewrite almost everything... Not sure that really makes sense and that this should be done. What would you gain from it? The idea was to have a lightweight Hatchlor for people using hatch (and there aren't that many) and PyScaffold for developers that want to stick with a traditional toolchain.

dornech committed:
fix: clean __init__.py to avoid Hatch dependency

dornech committed:
feat: make gitlint optional via cookiecutter variable

dornech commented:
Hm. My understanding was that PyScaffold helps to follow and adopt changes in an underlying template. I wonder why this should not be of interest for Hatch based delopment as for "traditional" toolchain. Only other tool I found is cruft but it is less well maintained and I am not yet too confident in this tool.

dornech committed:
feat: make gitlint optional via cookiecutter variable /2.

dornech committed:
feat: make gitlint optional via cookiecutter variable /2.

dornech committed:
fix: postgen_project.py hook for optional .gitlint

dornech committed:
fix: postgen_project.py hook for optional .gitlint /2.

dornech committed:
fix: hatch.version.raw-options according tohttps://github.com/ofek/hatch-vcs/discussions/12

FlorianWilhelm commented on {{cookiecutter.project_slug}}/pyproject.toml:
```diff
@@ -62,19 +69,117 @@ allow-direct-references = true
     "/.github",
 ]
 
+[tool.bumpversion]
```
@dornech, how does bumpversion play together with setuptools_scm? The latter takes the info from git, bumpversion uses a different concept, right?

dornech replied:
I am not sure if this is relevant. I do not really use setuptools_scm any more and also removed it as dependency for the "footgun" as dicsussed earlier. Setuptools_scm does not work with hatch / pyproject.toml because it does not get the raw-options.
I use bump-my-version (with calling generate-changelog as hook) or semantic-release to bump the version and generating a changelog automatically (what setuptools_scm does not do as far as I understood). What I wanted to achieve ist that the changelog is generated locally and together with creating a version before pushing to github and that works with these two toolchain amendments. The advantage here is also that with conventional commits (enforced by gitlint) this really works automatically.
What one could do: optional cookiecutter to use one or the other toolchain amendment - i. e. bump-my-version/generate-changelog or semantic-release. I have no real picture which one is better, both are pretty long "on the market" and properly maintained. handling is also pretty straightforward.

maresb commented:
Hey @dornech and @FlorianWilhelm! It's great to see all the activity in this PR. PyScaffold was actually my intro some years back to Python packaging best practices. :star_struck:

Based on @dornech's very helpful feedback I recently revamped [hatch-vcs-footgun-example](https://github.com/maresb/hatch-vcs-footgun-example/).

One important technical point regarding the version determination logic: it shouldn't be done within `__init__.py` but rather in some submodule like `version.py`. Otherwise it risks introducing circular imports if `__version__` is imported from elsewhere. (I just added a brief explanation of the problem [here](https://github.com/maresb/hatch-vcs-footgun-example/?tab=readme-ov-file#importerror-cannot-import-name-__version__-from-partially-initialized-module--most-likely-due-to-a-circular-import)).

Another thing is that I think you give me too much credit for hatch-vcs-footgun-example. It's a hack, and I think it should only be used when there's a specific need. When creating something like a project template for everyone to use, usually less is more. So in my subjective opinion, I think the sweet spot may be to create a simple `version.py` using just the `importlib.metadata.version` method, but then add a comment like:

```python
from importlib.metadata import version


# To update the version number in the project metadata it's necessary to reinstall
# the package. In case you are working with an editable install and need up-to-date
# version numbers without reinstallation, check out the alternative methods suggested
# in <https://github.com/maresb/hatch-vcs-footgun-example>.
__version__ = version(__project__)
```

Finally, from the perspective of a maintainer, in my experience it can be really challenging to approve larger PRs like this one. It requires the maintainer to carve out a big block of time to read the whole thing on each review cycle. If you could instead make several small PRs, then it becomes much easier / faster / safer to check correctness and merge each one individually. (That's the reason I stopped early in https://github.com/pypa/hatch/pull/1870 before achieving anything substantial; it's already probably way too big.)

FlorianWilhelm commented:
Hi @maresb and thanks for pointing that out and you feedback. I totally agree.

Would you also mind @dornech to split that PR a bit up? You have some fantastic improvements in there, which I would like to just merge but others really need to be discussed. It's much simpler if every isolated feature or improvement has its own PR instead of mixing them in one huge PR. Thank you.

dornech commented:
Hi, I admit i am not yet too experienced with Git / GitHub and "reasonable" assigning of changes and commits to pull requests. So I need some time and probaly assistance to sort that out.

maresb commented:
PRs are effectively just branches with a tiny bit of GitHub (not git) infrastructure on top to support conversations. So just focus on branches for now. Opening a PR for a branch is easy. Understanding Git is the eternal struggle.

It looks like you're developing from your main branch. Since your main is different from Florian's main, that's basically blocking you from submitting other PRs, so you should probably clear that up first...

From your `main`, run `git checkout -b various-amendments` to create a duplicate of your `main`. Then run `git push` and follow the instructions to set the upstream. Then go on GitHub and open a draft PR from `various-amendments`. You should add a reference to this PR. As far as I know, you can't change the source branch of a PR, so that means you need to close this one, but you'll have all the conversation history indirectly through the link on the new draft PR.

Now you need to reset your `main` branch via a hard reset. WARNING: this will clobber any tracked changes, meaning existing files that have been modified; new files you created will be unaffected. If you have any such files, run `git stash` first. (Run `git stash pop` afterwards to restore the tracked changes.) Run `git reset --hard upstream/main`. Now your `main` branch will be in order.

Finally, we can look at splitting this PR. The easiest way is if you can split the PR from the beginning. Like if you want to submit the first two commits as a PR, then check out the hash of the 2nd commit `git checkout 709b655`. You'll get a "detached head" warning, simply meaning you're not on a branch. Run `git checkout -b second-commit-branch` to create one and avoid decapitation. Now you can push this branch and create a PR for just the first two commits.

The next easiest way is via cherry-picking. Run `git checkout main` to get a clean starting point. Then if you want to bring in individual commits, run `git cherry-pick commit-hash`. Unfortunately you might have to solve merge conflicts when cherry-picking stuff out-of-order.

The hardest is rebasing, which involves rewriting the history of a branch. I'm not going to go there right now. :joy: 

Let me know if you have any questions.

dornech committed:
fix: update PSR config due to new release

changes for PSR release 10.x

dornech committed:
fix: improve GitHub workflows - reduce unnecessary triggers

optimized triggering of GitHub actions

dornech committed:
fix: GitHub action labeler

correct raw/endraw-markers

dornech committed:
fix: updates

update pyproject.toml and pre.commit-config.yaml

dornech committed:
fix: updates /2.

update pyproject.toml

dornech committed:
fix: updates /3.

update .pre-commit-config.yaml tmeplate

dornech committed:
fix: updates /4.

update pyproject.toml template

dornech committed:
fix: updates /5.

update pyproject.toml template and cookiecutter.json
add .cruft.json

dornech committed:
fix: updates /6.

clean-ups to use cruft - update.gitignore and pyproject.toml

dornech committed:
feat: include new version of hatch-vcs-footgun

adjust __init__.py, and pyproject.toml, include version.py, details see https://github.com/maresb/hatch-vcs-footgun-example

dornech committed:
fix: adjust config for pre-commit hooks (excludes, stages)

fix exclude-clause for cookiecutter-affected yml-files, use default_stages and also default_install_hook_types (hook commit_msg is required for gitlint and commitizen but leads to multiple hook executions without correct stage assignment)

dornech committed:
build: updates from dependabot

update GitHub workflows checkout 4 -> 5, download-artifact 4 -> 5 and gh-action-sigstore-pyton 3.0.0 -> 3.0.1

dornech committed:
feat: add commitizen config

include commitizen config in pyproject.toml, include improved commitizen changelog template aligned with python-semantic-release format (i. e. include commit-URL), adjust gitlint config to be more flexible and identify commit-msg from commitizen version bump

dornech committed:
BREAKING CHANGE: new version

extended toolchain (selectable via cookiecutter) including commitizen as alternative to semantic-relaase, cruft and much more

dornech committed:
BREAKING CHANGE: new version

extended toolchain (selectable via cookiecutter) including commitizen as alternative to semantic-relaase, cruft and much more

dornech committed:
BREAKING CHANGE: new version

extended toolchain (selectable via cookiecutter) including commitizen as alternative to semantic-relaase, cruft and much more

dornech committed:
BREAKING CHANGE: new version

extended toolchain (selectable via cookiecutter) including commitizen as alternative to semantic-release, cruft and much more

dornech committed:
BREAKING CHANGE: new version

extended toolchain (selectable via cookiecutter) including commitizen as alternative to semantic-release, cruft and much more

dornech committed:
fix: update setup-python dependency

upgrade v5 -> v6

dornech committed:
fix: correct .pre-commit-config.yaml

rev commitizen must be correct (error from initial copy&paste)

dornech committed:
fix: adjustments for commitizen tool-chain

note: new parameter "incremental" in template is assumed to be provided in a future standard commitizen release
small change to changelog template

dornech committed:
BREAKING CHANGE: implement target OS platform parameter

considering Windows platform dependent dependencies properly requires adjustment to GitHub Actions to narrow platform. Documentation build must run on Windows and overcome well-known cairo library issue.

dornech committed:
fix: correct hatch script

scripts for hatch environment lint was wrong, correct printing mypy and ruff version in linting script

dornech committed:
feat: add GitHub action for testbuild of documentation

documentation generation sometimes does not work in spite of commit checks passed. Initiate test build for every push to identify problem before version bumping.

dornech committed:
build: update pyproject.toml and GitHub Action for tests - minimum Python version 3.10

increment minimum Python version for own packages from 3.9 to 3.10 as some packages like Selenium or PyWin32 require Python>=3.10, include Python 3.14 in GitHub Action for testing

dornech committed:
fix: upgrade GitHub Actions

upgrade GitHub Action build.yml according to dependabot PRs applied to own packages

dornech committed:
fix: update cookiecutter.json (missing target_os_platform parameter)

update cookiecutter.json, add missing target_os_platform parameter

dornech committed:
fix: correct hatchlor test case definition and yaml files

update test environments with new parameters according to extension of the toolchain, adjust test scripts to overcome issues

FlorianWilhelm commented:
Hi @dornech, it's so cool to see that you regularly submit improvements to our beloved Hatchlor 🌹.

Since I have not so much time anymore to work on it. Would you want to become a maintainer directly? That would allow you to merge your PRs directly and change `main`. What do you think? 😃

dornech committed:
build(deps): bump actions in GitHub workflows

actions/download-artifacts to v8
actions/upload-artifacts to v7
crazy-max/ghaction-github-labeler to 6.0.0

dornech committed:
fix: upgrade GitHub actions

dornech committed:
fix: upgrade GitHub actions

dornech committed:
fix: correction cookiecutter.json

target platform options were not correct

dornech committed:
fix: GitHub Action related fixes

solve problem with Cairo library and add test script for finding/verifying Cairo library, upgrades of external GitHub Action modules

dornech committed:
fix: correct test config

update GitHub Actions, correct target platform parameter

dornech committed:
feat: switch from mkdocs and mkdocs-theme material to properdocs with theme materialx

mkdocs and mkdocs-theme material are no longer supported, properdocs and materialx are drop-in replacements
fix cookiecutter jinja2-variable expressions in properdocs.yml (former mkdocs.yml) by quiting or other means for commit

dornech committed:
fix: test environment

adjust GitHub Action for test

dornech committed:
fix: upgrade GitHub actions

dornech committed:
fix: upgrade GitHub actions

dornech committed:
fix: fix lockfile script

resolve test error

dornech committed:
fix: fix lockfile script next try

resolve test error

dornech committed:
fix: correct commitizen config

configuration for change type order was wrong, must be based on full  description

dornech committed:
fix: correct test script

lock file support was not working

dornech committed:
fix: correct test script

lock file support was not working

dornech committed:
fix: suspend lockfile test case

test of lock file support was not working, for meantime test suspended

dornech committed:
fix: upgrade GitHub actions

dornech committed:
fix: update version.py

update  of version.py to align with the hatch-vcs-footgun template

dornech committed:
fix: upgrade GitHub actions

dornech committed:
fix: adjust minimum Python version to 3.11 and other fixes

reason EOL Python 3.10 and experiencing version upshift in increasing number of packages, fix includes relaxed format check for version.py and editorial changes

dornech committed:
fix: upgrade GitHub actions

dornech committed:
fix: adjust minimum Python version to 3.11 - leftover cookiecutter.json

dornech committed:
fix: optimize GitHub actions resource usage, avoid especially very expensive macOS-actions

execute 'build' and 'test' actions only for ubuntu-latest (or windows-latest for windows specific packages) because pure Python packages are tagged "all" anyway.
execute 'documentation testbuild' action only for non-release commits.
